Add codec borsh feature

DaviRain-Su opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #259 

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->


______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

DaviRain-Su committed:
add codec feature

DaviRain-Su committed:
update

plafer commented on crates/ibc/src/core/ics02_client/client_type.rs:
```diff
@@ -2,10 +2,29 @@ use crate::prelude::*;
 use serde_derive::{Deserialize, Serialize};
 
+#[cfg(not(feature = "codec"))]
+#[cfg(not(feature = "borsh"))]
```
We should use [cfg_attr](https://doc.rust-lang.org/reference/conditional-compilation.html#the-cfg_attr-attribute) instead.

```rust
#[cfg_attr(feature = "scale-codec", derive(codec::Encode, codec::Decode))]
#[cfg_attr(feature = "borsh", derive(BorshSerialize, BorshDeserialize))]
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct ClientId(pub String);
```
In a follow-up PR we should also put serde `Serialize`/`Deserialize` behind a feature flag

DaviRain-Su replied:
nice cool

plafer commented on crates/ibc/Cargo.toml:
```diff
@@ -46,6 +48,11 @@ derive_more = { version = "0.99.17", default-features = false, features = ["from
+## for codec encode or decode
+codec = { package = "parity-scale-codec", version = "3.0.0", default-features = false, features = ["derive"] }
+scale-info = { version = "2.1.2", default-features = false, features = ["derive"] }
+sp-runtime = {  version = "7.0.0", default-features = false }
```
Can we make these [optional dependencies](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies)? Something like:

```toml
[dependencies]
# ...
codec = {..., optional = true}
scale-info = {..., optional = true}
sp-runtime = {..., optional = true}

[features]
# ...
scale-codec = ["dep:codec", "dep:scale-info", "dep:sp-runtime"]
```

plafer commented on crates/ibc/Cargo.toml:
```diff
@@ -18,8 +18,10 @@ all-features = true
-std = ["flex-error/std", "flex-error/eyre_tracer", "ibc-proto/std", "clock"]
+std = ["flex-error/std", "flex-error/eyre_tracer", "ibc-proto/std", "clock", "codec/std", "scale-info/std", "sp-runtime/std"]
 clock = ["tendermint/clock", "time/std"]
+codec = []
```
As per other comments, this would be renamed to `scale-codec`

DaviRain-Su committed:
Add Borsh feature and use cfg_attr replace cfg

DaviRain-Su committed:
Merge branch 'main' into add-codec-borsh-feature

DaviRain-Su committed:
Add scale-codec and borsh to Timestamp

DaviRain-Su committed:
fmt code

DaviRain-Su committed:
Update timestamp.rs

DaviRain-Su committed:
Update timestamp.rs

DaviRain-Su committed:
Merge branch 'main' into add-codec-borsh-feature

DaviRain-Su committed:
Create 259-add-borsh-and-scale-codec.md

plafer commented on crates/ibc/Cargo.toml:
```diff
@@ -35,8 +35,17 @@ std = [
 clock = ["tendermint/clock", "time/std"]
+scale-info = []
+sp-runtime = []
+codec = []
```
I think this can be removed; it is now scale-codec (which should be renamed to `parity-scale-codec` - see next comment).

plafer replied:
I think the only features we need are:

- std
- parity-scale-codec
- borsh

plafer commented on crates/ibc/Cargo.toml:
```diff
@@ -63,6 +72,12 @@ derive_more = { version = "0.99.17", default-features = false, features = ["from
 primitive-types = { version = "0.12.0", default-features = false, features = ["serde_no_std"] }
 dyn-clone = "1.0.8"
+## for codec encode or decode
+codec = { package = "parity-scale-codec", version = "3.0.0", default-features = false, features = ["derive"], optional = true }
```
I think it's best to have this package as `parity-scale-codec`.

plafer reviewed (changes requested):
Once again, great work here.

This is just a preliminary review, will finish it later!

DaviRain-Su committed:
Export feature parity-scale-codec, and borsh

plafer commented on crates/ibc/Cargo.toml:
```diff
@@ -35,8 +35,13 @@ std = [
     "tendermint/std",
+    "parity-scale-codec/std",
+    "scale-info/std",
+    "borsh/std"
```
We should not have any of these here; this effectively enables all three dependencies by default.

plafer commented on crates/ibc/Cargo.toml:
```diff
@@ -35,8 +35,13 @@ std = [
+    "borsh/std"
 ]
 clock = ["tendermint/clock", "time/std"]
+parity-scale-codec = ["dep:parity-scale-codec", "scale-info"]
```
```suggestion
parity-scale-codec = ["dep:parity-scale-codec", "dep:scale-info"]
```

plafer commented on crates/ibc/src/core/ics04_channel/packet.rs:
```diff
@@ -52,6 +52,18 @@ impl core::fmt::Display for PacketMsgType {
     }
 }
 
+#[cfg_attr(
+    feature = "parity-scale-codec",
+    derive(
+        parity_scale_codec::Encode,
+        parity_scale_codec::Decode,
+        scale_info::TypeInfo
+    )
+)]
+#[cfg_attr(
+    feature = "borsh",
+    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
+)]
```
Could we also add this to all **types that derive `serde::Serialize/Deserialize`** in modules:
+ `core::ics02_client::trust_threshold`
+ `core::ics03_connection::connection`
+ `core::ics03_connection::version`
+ `core::ics04_channel::channel`
+ `core::ics04_channel::commitment`
+ `core::ics04_channel::version`
+ `core::ics26_routing` (`ModuleId`)
+ `crate::Signer` type

DaviRain-Su replied:
done

plafer reviewed (changes requested):
almost there 😃

DaviRain-Su committed:
Update parity-scale-codec implement

- Add parity-scale-codec and Borsh below type
- core::ics02_client::trust_threshold
- core::ics03_connection::connection
- core::ics03_connection::version
- core::ics04_channel::channel
- core::ics04_channel::commitment
- core::ics04_channel::version
- core::ics26_routing (ModuleId)
- crate::Signer type

DaviRain-Su committed:
Update connection.rs

plafer commented on crates/ibc/src/core/ics02_client/client_type.rs:
```diff
@@ -2,10 +2,34 @@ use crate::prelude::*;
 pub struct ClientType(String);
 
+#[cfg(feature = "parity-scale-codec")]
+impl parity_scale_codec::Encode for ClientType {
```
We can use the derive macro if we [enable the `full` feature](https://github.com/paritytech/parity-scale-codec/issues/307#issuecomment-1004633308).

plafer commented on crates/ibc/src/core/ics02_client/client_type.rs:
```diff
@@ -2,10 +2,34 @@ use crate::prelude::*;
+    }
+}
+#[cfg(feature = "parity-scale-codec")]
+impl parity_scale_codec::Decode for ClientType {
```
Same here; we can use the derive macro with the `full` feature.

plafer commented on crates/ibc/src/core/ics03_connection/version.rs:
```diff
@@ -19,6 +24,71 @@ pub struct Version {
     features: Vec<String>,
 }
 
+mod sealed {
```
Similarly here, we don't need this module anymore if we use the `full` feature.

Same comment applies to all other instances of this pattern (if any).

plafer commented on crates/ibc/src/timestamp.rs:
```diff
@@ -26,6 +26,68 @@ pub struct Timestamp {
+        let timestamp = if let Some(time) = self.time {
+            time.unix_timestamp_nanos()
+        } else {
+            0i128
```
What is this value? Can we add a comment and/or a constant variable, whichever is most appropriate?

plafer commented on crates/ibc/src/timestamp.rs:
```diff
@@ -26,6 +26,68 @@ pub struct Timestamp {
+            .composite(scale_info::build::Fields::named().field(|f| {
+                f.ty::<Option<i128>>()
+                    .name("time")
+                    .type_name("Option<i128>")
```
Can we add a comment here that `i128` is chosen before we represent the timestamp is nanoseconds, which is represented as a `i128` by `Time`?

plafer reviewed (changes requested):
Left a few improvements!

DaviRain-Su committed:
Merge branch 'main' into add-codec-borsh-feature

DaviRain-Su committed:
Open parity-scale-codec features full

DaviRain-Su committed:
Update timestamp.rs

plafer commented on crates/ibc/src/core/ics03_connection/connection.rs:
```diff
@@ -97,6 +109,125 @@ pub struct ConnectionEnd {
     delay_period: Duration,
 }
 
+mod sealed {
```
This whole module can be removed by annotating `ConnectionEnd` directly (with `cfg_attr`), except for the `TypeInfo` implementation on line 199.

DaviRain-Su replied:
> 

Oh, it doesn't work here, `borsh` and `partiy-scale-code` don't automatically implement Encode and Decode for Duration, you have to do it manually.

plafer replied:
Ah yes, you're right; however `parity_scale_codec::Encode` and `parity_scale_codec::Decode` can be derived automatically.

```diff
diff --git a/crates/ibc/src/core/ics03_connection/connection.rs b/crates/ibc/src/core/ics03_connection/connection.rs
index c8717659..a52aab28 100644
--- a/crates/ibc/src/core/ics03_connection/connection.rs
+++ b/crates/ibc/src/core/ics03_connection/connection.rs
@@ -100,6 +100,13 @@ impl From<IdentifiedConnectionEnd> for RawIdentifiedConnection {
     }
 }
 
+#[cfg_attr(
+    feature = "parity-scale-codec",
+    derive(
+        parity_scale_codec::Encode,
+        parity_scale_codec::Decode,
+    )
+)]
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub struct ConnectionEnd {
     pub state: State,
@@ -112,14 +119,6 @@ pub struct ConnectionEnd {
 mod sealed {
     use super::*;
 
-    #[cfg_attr(
-        feature = "parity-scale-codec",
-        derive(
-            parity_scale_codec::Encode,
-            parity_scale_codec::Decode,
-            scale_info::TypeInfo
-        )
-    )]
     #[cfg_attr(
         feature = "borsh",
         derive(borsh::BorshSerialize, borsh::BorshDeserialize)
@@ -177,24 +176,6 @@ mod sealed {
         }
     }
 
-    #[cfg(feature = "parity-scale-codec")]
-    impl parity_scale_codec::Encode for ConnectionEnd {
-        fn encode_to<T: parity_scale_codec::Output + ?Sized>(&self, writer: &mut T) {
-            let value = InnerConnectionEnd::from(self.clone());
-            value.encode_to(writer);
-        }
-    }
-
-    #[cfg(feature = "parity-scale-codec")]
-    impl parity_scale_codec::Decode for ConnectionEnd {
-        fn decode<I: parity_scale_codec::Input>(
-            input: &mut I,
-        ) -> Result<Self, parity_scale_codec::Error> {
-            let result = InnerConnectionEnd::decode(input)?;
-            Ok(ConnectionEnd::from(result))
-        }
-    }
-
     #[cfg(feature = "parity-scale-codec")]
     impl scale_info::TypeInfo for ConnectionEnd {
         type Identity = Self;
```

plafer replied:
We should define our own wrapper type for `time::Duration`, and implement the `borsh` `Serialize`/`Deserialize` on it. We can do it in another PR though.

Created #302

DaviRain-Su replied:
> Ah yes, you're right; however parity_scale_codec::Encode and parity_scale_codec::Decode can be derived automatically.

Done

plafer commented on crates/ibc/src/core/ics24_host/identifier.rs:
```diff
@@ -162,9 +174,33 @@ impl From<String> for ChainId {
 pub struct ClientId(String);
 
+#[cfg(feature = "parity-scale-codec")]
+impl parity_scale_codec::Encode for ClientId {
```
We can use the derive macros on `ClientId` as well

plafer commented on crates/ibc/src/core/ics24_host/identifier.rs:
```diff
@@ -229,9 +265,33 @@ impl PartialEq<str> for ClientId {
 pub struct ConnectionId(String);
 
+#[cfg(feature = "parity-scale-codec")]
+impl parity_scale_codec::Encode for ConnectionId {
```
same here

plafer commented on crates/ibc/src/core/ics24_host/identifier.rs:
```diff
@@ -299,9 +359,33 @@ impl PartialEq<str> for ConnectionId {
 
+#[cfg(feature = "parity-scale-codec")]
+impl parity_scale_codec::Encode for PortId {
+    fn encode_to<T: parity_scale_codec::Output + ?Sized>(&self, writer: &mut T) {
```
and here

plafer commented on crates/ibc/src/core/ics24_host/identifier.rs:
```diff
@@ -346,9 +430,33 @@ impl Default for PortId {
 pub struct ChannelId(String);
 
+#[cfg(feature = "parity-scale-codec")]
+impl parity_scale_codec::Encode for ChannelId {
```
and here

plafer commented on crates/ibc/src/timestamp.rs:
```diff
@@ -26,6 +26,70 @@ pub struct Timestamp {
+        let timestamp = if let Some(time) = self.time {
+            time.unix_timestamp_nanos()
+        } else {
+            // CONTANT: When the value in time is a none we give the timestamp a default value of 0
```
```suggestion
            // When the value in `Time` is `None` we give the timestamp a default value of 0
```

plafer replied:
and this one

DaviRain-Su replied:
> and this one

done

plafer commented on crates/ibc/src/timestamp.rs:
```diff
@@ -26,6 +26,70 @@ pub struct Timestamp {
+            time.unix_timestamp_nanos()
+        } else {
+            // CONTANT: When the value in time is a none we give the timestamp a default value of 0
+            0i128
```
```suggestion
            0
```

plafer replied:
and this one

DaviRain-Su replied:
> and this one

done

plafer commented on crates/ibc/src/timestamp.rs:
```diff
@@ -26,6 +26,70 @@ pub struct Timestamp {
+        let timestamp = if let Some(time) = self.time {
+            time.unix_timestamp_nanos()
+        } else {
+            0i128
```
```suggestion
            0
```

plafer replied:
Can we also add the comment here:

> When the value in `Time` is `None` we give the timestamp a default value of 0

plafer replied:
Can we address these 2 points too please? @DaviRain-Su

DaviRain-Su replied:
done

DaviRain-Su replied:
> Can we address these 2 points too please? @DaviRain-Su

Dude, seriously, you're looking really hard.

plafer reviewed (changes requested):
A few uses of the derive macros were forgotten, and left a few nits!

DaviRain-Su committed:
Use Deive parity-scale-codec for ClientId, ConntctionId, PortId, ChannelId

DaviRain-Su committed:
Update connection.rs

DaviRain-Su committed:
Fix clippy

DaviRain-Su committed:
Update timestamp.rs

plafer reviewed (approved):
Thank you @DaviRain-Su! I very much appreciate your contributions 😄
